### PR TITLE
Add `ColumnarValue::values_to_arrays`, deprecate `columnar_values_to_array`

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/issues/8045 

## Rationale for this change
While working on implementing `ArrayToString` in  https://github.com/apache/arrow-datafusion/pull/9113 I found that many of the array functions support arbitrary combinations of arguments that are `ColumnarValue` to `Array` by converting all the arguments to arrays and then calling a function that works on arrays. 

I think this is a common desire when implementing functions, but it is not always clear how to do this.  

For example I think @Omega359  had to implement something similar in `to_date` in https://github.com/apache/arrow-datafusion/pull/9040

@viirya added a helper in https://github.com/apache/arrow-datafusion/pull/8962 which improves the situation, but I think this function may be hard to find (it is in `datafusion_physical_expr`) and it doesn't handle the case where the arguments are a mix of `ColumnarValue::Scalar` and `ColumnarValue::Array`

## What changes are included in this PR?
1. Move logic to `ColumnarValue::values_to_arrays`
2. Support mixed Scalar + ColumnarValues
3. Add documentation
4. Add unit tests
5. Deprecate the old function `columnar_values_to_array`

## Are these changes tested?

Yes, new tests

## Are there any user-facing changes?

New function, better docs and tests

### CODE REVIEW

File: datafusion-examples/examples/simple_udf.rs [hunk 1]
```diff
@@ -71,13 +70,15 @@ async fn main() -> Result<()> {
         // this is guaranteed by DataFusion based on the function's signature.
         assert_eq!(args.len(), 2);
 
-        let args = columnar_values_to_array(args)?;
+        // Expand the arguments to arrays (this is simple, but inefficient for
+        // single constant values).
+        let args = ColumnarValue::values_to_arrays(args)?;
```
Comment (alamb) [inline datafusion-examples/examples/simple_udf.rs:75]: This is the basic change -- make the logic a function of `ColumnarValue`

File: datafusion/expr/src/columnar_value.rs [hunk 2]
```diff
@@ -75,4 +75,166 @@ impl ColumnarValue {
...
+        Ok(args)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn values_to_arrays() {
```
Comment (alamb) [inline datafusion/expr/src/columnar_value.rs:132]: new tests

File: datafusion/expr/src/columnar_value.rs [hunk 3]
```diff
@@ -75,4 +75,166 @@ impl ColumnarValue {
...
+    /// Converts  [`ColumnarValue`]s to [`ArrayRef`]s with the same length.
+    ///
+    /// # Performance Note
+    ///
+    /// This function expands any [`ScalarValue`] to an array. This expansion
+    /// permits using a single function in terms of arrays, but it can be
+    /// inefficient compared to handling the scalar value directly.
+    ///
+    /// Thus, It is recommended to provide specialized implementations for
+    /// scalar values if performance is a concern.
```
Comment (viirya) [inline datafusion/expr/src/columnar_value.rs:88]:
```suggestion
    /// Thus, it is recommended to provide specialized implementations for
    /// scalar values if performance is a concern.
```

File: datafusion/expr/src/columnar_value.rs [hunk 4]
```diff
@@ -75,4 +75,166 @@ impl ColumnarValue {
...
+        for arg in args {
+            array_len = match (arg, array_len) {
+                (ColumnarValue::Array(a), None) => Some(a.len()),
+                (ColumnarValue::Array(a), Some(array_len)) => {
+                    if array_len == a.len() {
+                        Some(array_len)
+                    } else {
+                        return internal_err!(
+                            "Arguments has mixed length. Expected length: {array_len}, found length: {}",a.len()
+                        );
```
Comment (viirya) [inline datafusion/expr/src/columnar_value.rs:108]:
```suggestion
                            "Arguments has mixed length. Expected length: {array_len}, found length: {}", a.len()
                        );
```

**Review by viirya — APPROVED** [review]
Thank you @alamb. Looks good to me.

File: datafusion/expr/src/columnar_value.rs (hunk 4 above)
Comment (alamb) [reply datafusion/expr/src/columnar_value.rs:108]: That is strange fmt didn't pick that one up 🤔

**Comment by alamb** [comment]
Thanks again for the review @viirya

**Pull request merged**